<a href="https://colab.research.google.com/github/glorychern/ui-ux-learning-resource/blob/main/Copy_of_OSL_example_with_notes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

We start with a full 4 by 4 Hamiltonian $H_{full}$,

$$ H_{\text{full}} = \begin{bmatrix}
1.44792484 & 0.88504728 & 0.25784263 & 0.10590017 \\
0.88504728 & 3.67229552 & -0.26984386 & -0.01791287 \\
0.25784263 & -0.26984386 & 2.09992258 & -0.45397867 \\
0.10590017 & -0.01791287 & -0.45397867 & 2.77985706 \\
\end{bmatrix} $$

Suppose that $H_{\text{full}}$ is calculated in the following basis,
$$
\vert b_1 \rangle, \vert b_2 \rangle, \vert b_3 \rangle, \vert b_4 \rangle.
$$
and $$H_{\text{full}}(i,j) = \langle b_i \vert b_j \rangle$. Additionally suppose that we identified that $\vert b_1 \rangle, \vert b_3 \rangle$ were the two basis states with most significant contribution to the states we are interested in.

We then would like to find the effective Hamiltonian through OSL transformation.

We first diagonalize this $H_{\text{full}}$,

In [ ]:
import numpy as np

# Define the matrix
H_full = np.array([[1.44792484, 0.88504728, 0.25784263, 0.10590017],
                   [0.88504728, 3.67229552, -0.26984386, -0.01791287],
                   [0.25784263, -0.26984386, 2.09992258, -0.45397867],
                   [0.10590017, -0.01791287, -0.45397867, 2.77985706]])

# Perform eigendecomposition
eigenvalues, eigenvectors = np.linalg.eig(H_full)

# Create the diagonal matrix
H_d = np.diag(eigenvalues)

# Print the diagonalized matrix H_d
print("Diagonalized matrix H_d:")
print(H_d)

# Print the eigenvectors
print("Eigenvectors:")
print(eigenvectors)


Diagonalized matrix H_d:
[[1.00000001 0.         0.         0.        ]
 [0.         4.         0.         0.        ]
 [0.         0.         2.         0.        ]
 [0.         0.         0.         3.        ]]
Eigenvectors:
[[-0.87048568  0.3181021  -0.37117851 -0.0573782 ]
 [ 0.32381352  0.94101319  0.05916005 -0.07835214]
 [ 0.34225192 -0.10291829 -0.82243348 -0.44257724]
 [ 0.14234856  0.05208695 -0.4270009   0.89145615]]


In [ ]:
for i in range(len(eigenvalues)):
    eigenvector = eigenvectors[:, i]
    check = np.allclose(H_full @ eigenvector, eigenvalues[i] * eigenvector)
    print(f"Eigenvalue {i+1}: {eigenvalues[i]}")
    print(f"Eigenvector {i+1}: {eigenvector}")
    print(f"Explicit check for Eigenvalue {i+1}: {check}")


Eigenvalue 1: 1.0000000057535772
Eigenvector 1: [-0.87048568  0.32381352  0.34225192  0.14234856]
Explicit check for Eigenvalue 1: True
Eigenvalue 2: 3.9999999987355315
Eigenvector 2: [ 0.3181021   0.94101319 -0.10291829  0.05208695]
Explicit check for Eigenvalue 2: True
Eigenvalue 3: 1.999999996771928
Eigenvector 3: [-0.37117851  0.05916005 -0.82243348 -0.4270009 ]
Explicit check for Eigenvalue 3: True
Eigenvalue 4: 2.9999999987389665
Eigenvector 4: [-0.0573782  -0.07835214 -0.44257724  0.89145615]
Explicit check for Eigenvalue 4: True


We have obtained 4 eigen values with corresponding eigen vectors. Let us also demonstrate that the eigenvectors give us the unitary transform matrix $U_{full}$, we can check that $U_{full}$ is a unitary matrix,

In [ ]:
import numpy as np

# Given eigenvectors
eigenvectors = np.array([
    [-0.87048568, 0.32381352, 0.34225192, 0.14234856],
    [-0.37117851, 0.05916005, -0.82243348, -0.4270009],
    [-0.0573782, -0.07835214, -0.44257724, 0.89145615],
    [0.3181021, 0.94101319, -0.10291829, 0.05208695]
])

# Calculate the transpose of the matrix
transposed_matrix = eigenvectors.T

result=eigenvectors@transposed_matrix

result


array([[ 1.00000000e+00,  8.76627174e-10,  8.30858643e-09,
        -9.43592394e-09],
       [ 8.76627174e-10,  9.99999995e-01, -4.36064795e-10,
        -3.26271731e-09],
       [ 8.30858643e-09, -4.36064795e-10,  9.99999996e-01,
         1.52701550e-09],
       [-9.43592394e-09, -3.26271731e-09,  1.52701550e-09,
         9.99999995e-01]])

In [ ]:
# Check if the matrix is unitary
is_unitary = np.allclose(result, np.eye(4))

is_unitary

True

In [ ]:
# Define U_full as the eigenvectors
U_full = eigenvectors

# Calculate the inverse of U_full
U_full_inverse = np.linalg.inv(U_full)

# Demonstrate that H_d = U_full_inverse @ H_full @ U_full
result = np.allclose(H_d, U_full_inverse @ H_full @ U_full)
print("U_full_inverse @ H_full @ U_full:", U_full_inverse @ H_full @ U_full)
print("H_d = U_full_inverse @ H_full @ U_full:", result)


U_full_inverse @ H_full @ U_full: [[ 2.43965175  0.17260181  1.29663212  0.37935349]
 [ 0.17260182  2.79196156 -0.25662561 -0.45893598]
 [ 1.29663214 -0.2566256   2.26965162  0.54552086]
 [ 0.37935349 -0.45893597  0.54552086  2.49873508]]
H_d = U_full_inverse @ H_full @ U_full: False


At this point, our task is to identify which 2 out of the 4 eigenstate received most significant contribution from the two basis states $\vert b_1 \rangle, \vert b_3 \rangle$.

In [ ]:
import numpy as np

eigenvalues = np.array([1.0000000057535772, 3.9999999987355315, 1.999999996771928, 2.9999999987389665])
eigenvectors = np.array([[-0.87048568, 0.32381352, 0.34225192, 0.14234856],
                         [0.3181021, 0.94101319, -0.10291829, 0.05208695],
                         [-0.37117851, 0.05916005, -0.82243348, -0.4270009],
                         [-0.0573782, -0.07835214, -0.44257724, 0.89145615]])

# Calculate the square sum v[0]^2 + v[2]^2 for each eigenvector
square_sums = np.sum(eigenvectors[:, [0, 2]]**2, axis=1)

# Create a class to store eigenvalue, eigenvector, and square sum
class EigenClass:
    def __init__(self, eigenvalue, eigenvector, square_sum):
        self.eigenvalue = eigenvalue
        self.eigenvector = eigenvector
        self.square_sum = square_sum

# Create a list of EigenClass instances
eigen_classes = [EigenClass(eigenvalue, eigenvector, square_sum) for eigenvalue, eigenvector, square_sum in zip(eigenvalues, eigenvectors, square_sums)]

# Sort the eigen classes by square sum from largest to smallest
sorted_eigen_classes = sorted(eigen_classes, key=lambda x: x.square_sum, reverse=True)

# Print the sorted eigen classes
for eigen_class in sorted_eigen_classes:
    print(f"Eigenvalue: {eigen_class.eigenvalue}")
    print(f"Eigenvector: {eigen_class.eigenvector}")
    print(f"Square Sum: {eigen_class.square_sum}")
    print()


Eigenvalue: 1.0000000057535772
Eigenvector: [-0.87048568  0.32381352  0.34225192  0.14234856]
Square Sum: 0.8748816958287489

Eigenvalue: 1.999999996771928
Eigenvector: [-0.37117851  0.05916005 -0.82243348 -0.4270009 ]
Square Sum: 0.8141703153107306

Eigenvalue: 2.9999999987389665
Eigenvector: [-0.0573782  -0.07835214 -0.44257724  0.89145615]
Square Sum: 0.1991668712012576

Eigenvalue: 3.9999999987355315
Eigenvector: [ 0.3181021   0.94101319 -0.10291829  0.05208695]
Square Sum: 0.1117811204409341



As we have now identified that eigenstates with eigenvalues $1$ and $2$ receives most significant contributions from the two basis states $\vert b_1 \rangle, \vert b_3 \rangle$. So we rearrange $H_d$ and swap column of $U_{full}$ accordingly.

$$ H_{\text{d}(new)} = \begin{bmatrix}
1.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 2.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 4.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 3.0 \\
\end{bmatrix} $$

and the corresponding $U_{full(new)}$ as,

$$
U_{full(new)}=\begin{bmatrix}
-0.87048568 & -0.37117851 & 0.3181021 & -0.0573782 \\
0.32381352 & 0.05916005 & 0.94101319 & -0.07835214 \\
0.34225192 & -0.82243348 & -0.10291829 & -0.44257724 \\
0.14234856 & -0.4270009 & 0.05208695 & 0.89145615 \\
\end{bmatrix}
$$




In [ ]:
import numpy as np

H_d_new = np.diag([1, 2, 4, 3])

U_full_new = np.array([[-0.87048568, -0.37117851, 0.3181021, -0.0573782],
                      [0.32381352, 0.05916005, 0.94101319, -0.07835214],
                      [0.34225192, -0.82243348, -0.10291829, -0.44257724],
                      [0.14234856, -0.4270009, 0.05208695, 0.89145615]])

U_full_new_inverse = np.linalg.inv(U_full_new)

result = U_full_new_inverse @ H_full @ U_full_new
check = np.allclose(result, H_d_new, atol=1e-6)

print("U_full_new_inverse @ H_full @ U_full_new:")
print(result)
print("H_d_new:")
print(H_d_new)
print("Check if U_full_new_inverse @ H_full @ U_full_new = H_d_new:")
print(check)


U_full_new_inverse @ H_full @ U_full_new:
[[ 1.00000001e+00  1.59240233e-09  1.10127973e-08 -7.41444731e-09]
 [ 2.46902954e-09  2.00000000e+00 -2.29922611e-09 -3.84579391e-09]
 [-1.72949743e-08 -8.82466077e-09  4.00000000e+00  2.81398621e-09]
 [ 9.20272536e-09 -4.28185865e-09  1.28697059e-09  3.00000000e+00]]
H_d_new:
[[1 0 0 0]
 [0 2 0 0]
 [0 0 4 0]
 [0 0 0 3]]
Check if U_full_new_inverse @ H_full @ U_full_new = H_d_new:
True


As of now, we identified our $(H_d)_P$ as

$$
(H_d)_P = \begin{bmatrix}
1.0 & 0 \\
0 & 2.0 \\
\end{bmatrix}
$$


$U_P$ as a 2 by 2 matrix,

$$
U_P = \begin{bmatrix}
-0.87048568 & -0.37117851 \\
0.32381352 & 0.05916005 \\
\end{bmatrix}
$$

We need to unitarized $U_P$,

$$
U_{P(unitary)} = \frac{U_P}{\sqrt{U_P^\dagger U_P}} , \quad U_{P(unitary)}^\dagger = \frac{U_P^\dagger}{\sqrt{U_P^\dagger U_P}}
$$

Practically, we calculate the as following,

$$
U_{P(unitary)} = \sqrt{U_P U_P^\dagger} U_P^{\dagger-1} , \quad U_{P(unitary)}^\dagger = (U_P)^{-1}\sqrt{U_P U_P^\dagger}
$$

In [ ]:
import numpy as np
from scipy.linalg import sqrtm

# Define U_P matrix
U_P = np.array([[-0.87048568, -0.37117851],
                [0.32381352, 0.05916005]])

# Transpose of U_P
U_P_T = U_P.T

# U_P times Transpose of U_P
result = U_P @ U_P_T

# Square root of the result
sqrt_result = sqrtm(result)

# You can check that sqrt_result@sqrt_result = result

# Element-wise division of U_P by the square root of the result
U_P_unitary = sqrt_result@np.linalg.inv(U_P_T)
U_P_T_unitary= np.linalg.inv(U_P)@sqrt_result

# Display the results
print("U_P divided by the square root of U_P times U_P_T:")
print(U_P_unitary)
print("U_P_T divided by the square root of U_P times U_P_T:")
print(U_P_T_unitary)


U_P divided by the square root of U_P times U_P_T:
[[-0.75945543 -0.65055934]
 [ 0.65055934 -0.75945543]]
U_P_T divided by the square root of U_P times U_P_T:
[[-0.75945543  0.65055934]
 [-0.65055934 -0.75945543]]


In [ ]:
U_P_unitary@U_P_T_unitary

array([[1.00000000e+00, 5.55111512e-16],
       [7.21644966e-16, 1.00000000e+00]])

In [ ]:
# Check if division_result is unitary
is_unitary = np.allclose(np.eye(2), U_P_unitary@U_P_unitary.T)

# Display the result
print("Is division_result unitary?", is_unitary)

Is division_result unitary? True


So now we have found out unitary matrix in P space,


$$
U_{P(unitary)} = \begin{bmatrix}
-0.75945543 & -0.65055934 \\
0.65055934 & -0.75945543
\end{bmatrix}
$$
and
$$
U_{P(unitary)}^\dagger =
\begin{bmatrix}
-0.75945543 & 0.65055934 \\
-0.65055934 & -0.75945543 \\
\end{bmatrix}
$$
and our effective Hamiltonian after OSL transformation is,
$$
H_{eff} = U_{P(unitary)}^\dagger (H_d)_P U_{P(unitary)}
$$
The numerical result is,
$$H_{eff} =
\begin{bmatrix}
1.42322746 & -0.49407082 \\
-0.49407082 & 1.57677254 \\
\end{bmatrix}
$$



In [ ]:
U_P_T_unitary@np.diag(np.arange(1, 3))@U_P_unitary

array([[ 1.42322746, -0.49407082],
       [-0.49407082,  1.57677254]])

In [ ]:
import numpy as np

# Define the matrix
matrix = np.array([[1.42322746, -0.49407082],
                   [-0.49407082, 1.57677254]])

# Perform eigendecomposition
eigenvalues, eigenvectors = np.linalg.eig(matrix)

# Create the diagonal matrix
diagonal_matrix = np.diag(eigenvalues)

# Print the diagonalized matrix
print("Diagonalized matrix:")
print(diagonal_matrix)

# Print the eigenvectors
print("Eigenvectors:")
print(eigenvectors)


Diagonalized matrix:
[[1. 0.]
 [0. 2.]]
Eigenvectors:
[[-0.75945542  0.65055934]
 [-0.65055934 -0.75945542]]
